lib/styles/trackInputModality.js

```diff
@@ -33,25 +88,39 @@ function setUpEventHandlers(disableFocusRingByDefault) {
 
   document.body.addEventListener(
     'keydown',
-    event => {
-      recentKeyboardEvent = event;
+    () => {
+      hadKeyboardEvent = true;
+      hadClickEvent = false;
 
       if (isHandlingKeyboardThrottle) {
         clearTimeout(isHandlingKeyboardThrottle);
       }
 
       isHandlingKeyboardThrottle = setTimeout(() => {
-        recentKeyboardEvent = null;
+        hadKeyboardEvent = false;
       }, 100);
     },
     true
   );
 
+  document.body.addEventListener('mouseup', () => {
+    hadClickEvent = true;
+    hadKeyboardEvent = false;
+  });
+
   document.body.addEventListener(
     'focus',
-    () => {
-      if (recentKeyboardEvent && KEYS_WHITELIST.includes(recentKeyboardEvent.key)) {
+    e => {
+      if (
+        (hadKeyboardEvent && focusTriggersKeyboardModality(e.target)) ||
+        (focusSetExplicitly(e.target) && !hadClickEvent)
+      ) {
+        // both the JS state and the body attribute for keyboard modality
         globalThemeState.inputModality = 'keyboard';
+        document.body.setAttribute('modality', 'keyboard');
+      } else {
+        globalThemeState.inputModality = null;
+        document.body.setAttribute('modality', '');
```

Comment:
We need to also remove the `modality` attribute from the body element in the `blur` listener below, otherwise, we will only clear it when there are focus events after a mouse click.